opencog/atoms/core/ImplicationLink.cc

```diff
@@ -61,26 +61,23 @@ ImplicationLink::ImplicationLink(Link &l)
 void ImplicationLink::extract_variables(const HandleSeq& oset)
 {
 	size_t sz = _outgoing.size();
-	// For now we assume no implicit scoping, so the number of
-	// arguments is 2 when sugar syntax isn't used (no scoping), or 3
-	// when the sugar syntax (scoping) is used.
+
 	if (sz < 2 or 3 < sz)
 		throw InvalidParamException(TRACE_INFO,
-			"Expecting an outgoing set size of at most two, got %d", sz);
+			"Expecting an outgoing set of size 2 or 3, got %d", sz);
+
+	// If there is an explicit variable declaration, then use that.
+	// If there is no variable declaration, then assume that all of
+	// the free variables in the link are scoped.
+	ScopeLink::extract_variables(oset);
```

Comment:
@linas this is wrong. In the following

```atomese
Implication
  Variable "$X"
  Predicate "P"
```

you don't want `$X` to become a scoped variable under the `Implication`.

Due to that the following PutLink fails

```scheme
(PutLink
   (LambdaLink
      (VariableList
         (TypedVariableLink
            (VariableNode "$X")
            (TypeNode "PredicateNode")
         )
         (TypedVariableLink
            (VariableNode "$Y")
            (TypeNode "PredicateNode")
         )
      )
      (AndLink
         (ImplicationLink
            (VariableNode "$X")
            (PredicateNode "is-human")
         )
         (ImplicationLink
            (VariableNode "$Y")
            (PredicateNode "is-human")
         )
         (EvaluationLink
            (PredicateNode "acquainted")
            (ListLink
               (VariableNode "$X")
               (VariableNode "$Y")
            )
         )
      )
   )
   (ListLink
      (PredicateNode "is-Self")
      (PredicateNode "is-Bob")
   )
)
```

This is a perfect example as to why implicit scoping is a bad idea (unless regulated by a set of rules allowing complete unambiguous expressiveness, which is not the case ATM).